File: frontend/js/vessels.js
// Ships, drawn from their own broadcasts.
//
// AIS is a VHF transmission every vessel over a certain size makes
// continuously: who it is, where it is, how fast and which way. Receivers
// ashore hear it and some publish what they hear.
//
// Which makes it a natural third thing to have beside the imagery. A Sentinel-1
// pass shows a bright dot on the water and cannot tell you what it is; AIS
// names the ships that chose to be named. Where a dot has no ship against it,
// that is worth noticing on its own.

import { store } from './store.js';
import { $, el, toast, debounce } from './ui.js';
import { api } from './api.js';

// Colour by what the ship is for. Tankers and cargo are the two anyone is
// really looking for, so they get the colours that carry furthest.
const CATEGORIES = {
  tanker: { colour: '#ff8f5e', label: 'Tanker' },
  cargo: { colour: '#4cc2ff', label: 'Cargo' },
  passenger: { colour: '#3ee39f', label: 'Passenger' },
  fast: { colour: '#c78bff', label: 'High-speed' },
  special: { colour: '#ffd166', label: 'Fishing, tug, pilot' },
  wig: { colour: '#9aa7bd', label: 'Wing-in-ground' },
  other: { colour: '#8b98ad', label: 'Other' },
};

// Below this a ship is drawn as a dot rather than an arrow: a vessel that is
// not moving has no direction worth pointing, and a moored fleet of arrows all
// aimed at whatever heading they last reported reads as motion that is not
// happening.
const MOVING_KNOTS = 0.6;

// Panning inside the area already fetched should not fetch again. The pad is
// how much slack there is before it does.
const PAD = 0.25;

// Where the ships come from.
//
// Digitraffic is open and asks for nothing, and covers the Baltic. aisstream
// covers the world and wants a free API key. Neither is better; they answer
// different questions, and the panel says which is which rather than leaving
// an empty map to be interpreted.
const SOURCES = {
  digitraffic: {
    label: 'Baltic', needsKey: false,
    reach: 'Baltic & Gulf of Finland',
    note: 'Digitraffic — open, no key needed.',
  },
  aisstream: {
    label: 'Global', needsKey: true,
    reach: 'Anywhere, with a key',
    note: 'aisstream.io — asks the service at most once every five minutes.',
  },
};

let map = null;
let layer = null;
let enabled = false;
let showNames = false;
let covered = null;      // last box fetched, padded
let ships = [];
let refresher = null;
let source = 'digitraffic';
// How long before the global feed may be asked again. The backend enforces
// this; the panel only reports it, so a reload cannot talk it into going
// sooner.
let nextIn = 0;
// Whether the backend is holding a key. Tracked here as well as in the config
// because the config is read once at startup and would still say "no key" for
// the rest of the session after one was pasted in.
let keySaved = false;

export function initVessels(leafletMap) {
  map = leafletMap;
  map.createPane('vessels').style.zIndex = 470;
  buildDock();
  map.on('moveend', debounce(() => { if (enabled) maybeRefetch(); }, 350));
}

// ── Fetching ───────────────────────────────────────────────────

const boxOf = (bounds) => [
  bounds.getWest(), bounds.getSouth(), bounds.getEast(), bounds.getNorth(),
];

function padded([w, s, e, n]) {
  const dx = (e - w) * PAD;
  const dy = (n - s) * PAD;
  return [w - dx, s - dy, e + dx, n + dy];
}

const inside = (box, outer) =>
  box[0] >= outer[0] && box[1] >= outer[1] && box[2] <= outer[2] && box[3] <= outer[3];

async function maybeRefetch() {
  const box = boxOf(map.getBounds());
  if (covered && inside(box, covered)) return;
  await load(box);
}

async function load(box) {
  const ask = padded(box);
  // The global feed listens to the stream for a few seconds before answering,
  // which is a long time to sit in front of a panel that says nothing.
  note(source === 'aisstream' ? 'Listening to the stream…' : 'Loading…');
  try {
    const data = await api.vessels(ask, source);
    covered = ask;
    nextIn = data.next_in ?? 0;
    ships = data.vessels ?? [];
    draw(data);
  } catch (err) {
    // The panel carries the detail; the toast only says something went wrong,
    // since the message can be a paragraph and a toast is not the place.
    toast('AIS: no ships loaded', 'err');
    note(err.message, 'bad');
  }
}

// ── Drawing ────────────────────────────────────────────────────

/**
 * A ship as an arrow pointing where it is going, or a dot if it is not going.
 *
 * Heading is where the bow points and course is where the ship is actually
 * travelling; they differ in a current or a crosswind. Heading is used when
 * the ship reports one, because that is what a shape on a map is showing.
 */
function marker(ship) {
  const spec = CATEGORIES[ship.category] ?? CATEGORIES.other;
  const moving = (ship.speed ?? 0) >= MOVING_KNOTS;
  const angle = ship.heading ?? ship.course ?? 0;

  const icon = moving
    ? L.divIcon({
      className: 'ship',
      html: `<svg viewBox="0 0 24 24" style="transform:rotate(${angle}deg)">
               <path d="M12 2 L18 21 L12 17 L6 21 Z" fill="${spec.colour}"
                     stroke="rgba(0,0,0,.55)" stroke-width="1.2"/></svg>`,
      iconSize: [18, 18], iconAnchor: [9, 9],
    })
    : L.divIcon({
      className: 'ship',
      html: `<svg viewBox="0 0 24 24"><circle cx="12" cy="12" r="6" fill="${spec.colour}"
               stroke="rgba(0,0,0,.55)" stroke-width="1.5"/></svg>`,
      iconSize: [13, 13], iconAnchor: [6.5, 6.5],
    });

  const pin = L.marker([ship.lat, ship.lon], {
    icon, pane: 'vessels', riseOnHover: true,
    title: ship.name || `MMSI ${ship.mmsi}`,
  });
  pin.bindPopup(() => describe(ship), { className: 'ship-popup', maxWidth: 260 });

  if (showNames && ship.name) {
    pin.bindTooltip(ship.name, {
      permanent: true, direction: 'right', offset: [10, 0], className: 'ship-name',
    });
  }
  return pin;
}

function describe(ship) {
  const spec = CATEGORIES[ship.category] ?? CATEGORIES.other;
  const rows = [
    ['Type', ship.type],
    ['Speed', ship.speed == null ? null : `${ship.speed} kn`],
    ['Course', ship.course == null ? null : `${Math.round(ship.course)}°`],
    ['Status', ship.status],
    ['Bound for', ship.destination],
    ['Draught', ship.draught ? `${ship.draught} m` : null],
    ['Length', ship.length ? `${ship.length} m` : null],
    ['MMSI', ship.mmsi],
    ['Call sign', ship.callsign],
    // How stale the position is. A ship that stopped broadcasting an hour ago
    // is not where this says it is, and the popup should not imply otherwise.
    ['Reported', ship.age_min == null ? null
      : ship.age_min < 1 ? 'just now' : `${Math.round(ship.age_min)} min ago`],
  ].filter(([, value]) => value != null && value !== '');

  return el('div', { class: 'ship-card' },
    el('div', { class: 'ship-head' },
      el('span', { class: 'ship-dot', style: `background:${spec.colour}` }),
      el('b', {}, ship.name || `MMSI ${ship.mmsi}`)),
    el('dl', {}, ...rows.flatMap(([label, value]) =>
      [el('dt', {}, label), el('dd', {}, String(value))])),
  ).outerHTML;
}

function draw(data) {
  layer?.remove();
  layer = L.layerGroup(ships.map(marker), { pane: 'vessels' });
  layer.addTo(map);

  if (data && !data.covered) {
    note(data.note ?? 'No AIS coverage here.', 'warn');
    return;
  }
  const shown = ships.length;
  const total = data?.count ?? shown;
  if (shown) {
    note(`${shown} vessel${shown === 1 ? '' : 's'}${total > shown ? ` of ${total}` : ''}`
      + `${data?.demo ? ' — synthetic' : ''}${data?.cached ? ' · from the last look' : ''}`);
  } else if (data?.messages === 0) {
    // Nothing arrived at all, which is a different problem from an empty sea
    // and wants a different answer from whoever is looking at it.
    note('Connected, but nothing came through. Try a wider area.', 'warn');
  } else {
    note('No vessels broadcasting here right now.', 'warn');
  }
  legend();
}

function note(text, kind = '') {
  const box = $('#vesselNote');
  if (!box) return;
  box.className = `vessel-status${kind ? ` is-${kind}` : ''}`;
  box.textContent = text;
}

function legend() {
  const host = $('#vesselKey');
  if (!host) return;
  const seen = new Set(ships.map((s) => s.category));
  host.replaceChildren(...Object.entries(CATEGORIES)
    .filter(([key]) => seen.has(key))
    .map(([, spec]) => el('span', { class: 'ship-swatch' },
      el('i', { style: `background:${spec.colour}` }), spec.label)));
}

// ── Turning it on and off ──────────────────────────────────────

async function toggle() {
  enabled = !enabled;
  paintDock();
  if (!enabled) {
    clearInterval(refresher);
    layer?.remove();
    layer = null;
    covered = null;
    ships = [];
    return;
  }
  await load(boxOf(map.getBounds()));
  // Half a minute suits the keyless Baltic feed, which is a plain HTTP
  // request against a cache. The global one has a five-minute floor of its
  // own in the backend, so asking more often would only be answered from
  // what it already has -- there is no point pretending otherwise.
  const every = source === 'aisstream' ? 5 * 60 * 1000 : 30000;
  refresher = setInterval(() => { covered = null; maybeRefetch(); }, every);
}

// ── The panel ──────────────────────────────────────────────────

async function setSource(next) {
  if (next === source) return;
  source = next;
  covered = null;
  clearInterval(refresher);
  buildDock();
  if (!enabled) return;
  await load(boxOf(map.getBounds()));
  const every = source === 'aisstream' ? 5 * 60 * 1000 : 30000;
  refresher = setInterval(() => { covered = null; maybeRefetch(); }, every);
}

async function saveKey(value) {
  try {
    const out = await api.aisKey(value);
    keySaved = out.set;
    buildDock();
    note(out.set ? 'Key saved. Listening…' : 'Key cleared.');
    if (out.set && enabled) { covered = null; await load(boxOf(map.getBounds())); }
  } catch (err) {
    note(err.message, 'bad');
  }
}

function buildDock() {
  const dock = $('#vesselDock');
  if (!dock) return;
  const spec = SOURCES[source];
  dock.innerHTML = '';
  dock.append(
    el('button', { class: 'vessel-toggle', id: 'vesselToggle', onclick: toggle },
      el('span', { class: 'vessel-mark' }, '⛴'), 'Ships'),
    el('div', { class: 'vessel-body', id: 'vesselBody', hidden: !enabled },

      // A segmented control rather than two buttons that happen to sit
      // together: one track, one lit segment, so which of the two is in force
      // reads at a glance instead of having to be compared.
      el('div', { class: 'seg', role: 'tablist' },
        ...Object.entries(SOURCES).map(([key, opt]) =>
          el('button', {
            class: `seg-item${key === source ? ' is-on' : ''}`,
            dataset: { source: key }, role: 'tab',
            'aria-selected': String(key === source),
            title: opt.reach,
            onclick: () => setSource(key),
          }, opt.label))),

      el('div', { class: 'vessel-reach', id: 'vesselReach' }, spec.reach),

      // The key, and only when the chosen feed wants one.
      ...(spec.needsKey ? [
        el('div', { class: 'keyfield', id: 'keyField' },
          el('input', {
            type: 'password', id: 'aisKey', autocomplete: 'off', spellcheck: false,
            placeholder: keySaved ? 'Key saved' : 'Paste your API key',
            onkeydown: (e) => { if (e.key === 'Enter') saveKey(e.target.value); },
            oninput: () => paintKeyButton(),
          }),
          el('button', {
            class: 'keyfield-go', id: 'aisKeySave', type: 'button',
            onclick: () => saveKey($('#aisKey').value),
          }, keySaved ? 'Change' : 'Save')),
        el('div', { class: 'vessel-hint' },
          keySaved ? el('span', { class: 'ok-dot' }, '● key saved') : '',
          el('a', {
            href: 'https://aisstream.io/', target: '_blank', rel: 'noopener noreferrer',
          }, 'Get a free key'),
          ' · kept in memory only'),
        // When the map is empty, guessing stops being useful. This connects
        // once over the Dover Strait -- busy at every hour -- and says what
        // actually happened.
        el('button', {
          class: 'vessel-test', id: 'aisTest', type: 'button',
          onclick: runTest,
        }, 'Test the connection'),
      ] : []),

      el('label', { class: 'vessel-check' },
        el('input', {
          type: 'checkbox', checked: showNames,
          onchange: (e) => { showNames = e.target.checked; draw(null); },
        }), 'Show names'),

      el('div', { class: 'vessel-key', id: 'vesselKey' }),
      el('div', { class: 'vessel-status', id: 'vesselNote' }, enabled ? 'Loading…' : ''),
      el('div', { class: 'vessel-where', id: 'vesselWhere' })),
  );
  // The button starts out with nothing to save, and should say so.
  paintKeyButton();
  paintDock();
}

async function runTest() {
  const button = $('#aisTest');
  if (button) { button.disabled = true; button.textContent = 'Testing…'; }
  note('Connecting over the Dover Strait…');
  try {
    const out = await api.aisTest();
    note(out.detail, out.ok ? '' : 'bad');
  } catch (err) {
    note(err.message, 'bad');
  } finally {
    if (button) { button.disabled = false; button.textContent = 'Test the connection'; }
  }
}

/** The save button only offers itself when there is something to save. */
function paintKeyButton() {
  const field = $('#aisKey');
  const button = $('#aisKeySave');
  if (!field || !button) return;
  button.disabled = !field.value.trim();
}

function paintDock() {
  const button = $('#vesselToggle');
  const body = $('#vesselBody');
  if (!button || !body) return;
  button.classList.toggle('is-on', enabled);
  body.hidden = !enabled;
  keySaved = keySaved || Boolean(store.config?.vessels?.global_key_set);
  const where = $('#vesselWhere');
  if (where && enabled) {
    // Say where the feed reaches, up front. A blank map over the Atlantic is
    // otherwise indistinguishable from a broken layer.
    const wait = nextIn > 0 && source === 'aisstream'
      ? ` Next look in ${Math.ceil(nextIn / 60)} min.` : '';
    where.textContent = SOURCES[source].note + wait;
  }
}

File: backend/vessels.py
"""Ships, from AIS.

Every vessel over a certain size broadcasts its position, course and identity
continuously on VHF. Receivers ashore pick it up and some of them publish it.

Getting that feed for free and without an account is the hard part, and worth
being straight about: there is no open global AIS service. The one used here is
Digitraffic, run by the Finnish Transport Infrastructure Agency, which is
genuinely open -- no key, no signup, documented and stable -- and covers the
Baltic and the Gulf of Finland. Outside that box it returns nothing, and the
app says so rather than looking broken.

A global feed means an account somewhere. The front end can take an aisstream.io
key for that; this module is the part that needs nothing.
"""

from __future__ import annotations

import datetime as dt
import math
import threading
import time

import requests

from . import config


class VesselLookupError(RuntimeError):
    pass


SOURCE = {
    "key": "digitraffic",
    "label": "Digitraffic (Finnish Transport Infrastructure Agency)",
    "locations": "https://meri.digitraffic.fi/api/ais/v1/locations",
    "vessels": "https://meri.digitraffic.fi/api/ais/v1/vessels",
    # Roughly what the network hears: the Baltic, the Gulf of Bothnia and the
    # Gulf of Finland. Checked before asking, so a map over the Pacific gets an
    # explanation instead of an empty answer.
    "bounds": (9.0, 53.0, 32.0, 66.5),
    "attribution": "Digitraffic / Fintraffic, CC BY 4.0",
}

# The feed updates every few seconds, but nothing is served from it faster than
# this: the map pans, and re-asking on every drag would be rude to a free
# service that asks for nothing in return.
CACHE_SECONDS = 20

# Vessel names change about never, so they are held far longer than positions
# and refreshed in the background rather than on the path of a request.
NAMES_SECONDS = 30 * 60

# AIS ship type is a two-digit code. The tens digit is the category, which is
# all anyone needs to tell a tanker from a tug on a map.
TYPES = {
    2: ("wig", "Wing-in-ground"),
    3: ("special", "Fishing, tug or dredger"),
    4: ("fast", "High-speed craft"),
    5: ("special", "Pilot, rescue or patrol"),
    6: ("passenger", "Passenger"),
    7: ("cargo", "Cargo"),
    8: ("tanker", "Tanker"),
    9: ("other", "Other"),
}

# What the navigational status field means. Under way is the ordinary case;
# the rest are worth showing because they explain a ship that is not moving.
NAV_STATUS = {
    0: "under way (engine)", 1: "at anchor", 2: "not under command",
    3: "restricted manoeuvrability", 4: "constrained by draught", 5: "moored",
    6: "aground", 7: "fishing", 8: "under way (sailing)",
    15: "undefined",
}

_lock = threading.Lock()
_positions: dict = {"at": 0.0, "data": None}
_names: dict = {"at": 0.0, "data": {}}

_session = requests.Session()
_session.headers["User-Agent"] = config.USER_AGENT
# Digitraffic asks callers to identify themselves, and it is a reasonable
# thing to ask of a service that charges nothing.
_session.headers["Digitraffic-User"] = "EarthViewer/1.0"


def _covered(bbox: tuple[float, float, float, float]) -> bool:
    west, south, east, north = bbox
    w, s, e, n = SOURCE["bounds"]
    return not (east < w or west > e or north < s or south > n)


def _fetch(url: str) -> dict:
    try:
        resp = _session.get(url, timeout=25)
        resp.raise_for_status()
    except requests.RequestException as exc:
        raise VesselLookupError(f"{SOURCE['label']} could not be reached: {exc}") from exc
    return resp.json()


def _all_positions() -> dict:
    """Every position the service is publishing, cached briefly."""
    now = time.time()
    with _lock:
        if _positions["data"] is not None and now - _positions["at"] < CACHE_SECONDS:
            return _positions["data"]
    data = _fetch(SOURCE["locations"])
    with _lock:
        _positions.update(at=now, data=data)
    return data


def _all_names() -> dict[int, dict]:
    """Names and types by MMSI, cached for a long time."""
    now = time.time()
    with _lock:
        if _names["data"] and now - _names["at"] < NAMES_SECONDS:
            return _names["data"]
    try:
        rows = _fetch(SOURCE["vessels"])
    except VesselLookupError:
        # A ship with no name is still a ship. Losing the lookup should not
        # lose the positions with it.
        return _names["data"]
    table = {}
    for row in rows if isinstance(rows, list) else []:
        mmsi = row.get("mmsi")
        if mmsi is None:
            continue
        table[int(mmsi)] = {
            "name": (row.get("name") or "").strip() or None,
            "callsign": (row.get("callSign") or "").strip() or None,
            "type": row.get("shipType"),
            "destination": (row.get("destination") or "").strip() or None,
            "draught": row.get("draught"),
            "length": (row.get("referencePointA") or 0) + (row.get("referencePointB") or 0) or None,
        }
    with _lock:
        _names.update(at=now, data=table)
    return table


def _kind(ship_type) -> tuple[str, str]:
    """AIS ship type code to a category anyone can read off a map."""
    try:
        tens = int(ship_type) // 10
    except (TypeError, ValueError):
        return "other", "Unknown"
    return TYPES.get(tens, ("other", "Other"))


def _age(stamp) -> float | None:
    """Minutes since the position was broadcast."""
    try:
        return max(0.0, (time.time() * 1000 - float(stamp)) / 60000.0)
    except (TypeError, ValueError):
        return None


def vessels_in(bbox: tuple[float, float, float, float], limit: int = 900) -> dict:
    """Every vessel inside a rectangle, nearest the centre first."""
    west, south, east, north = bbox
    if not _covered(bbox):
        return {
            "vessels": [], "count": 0, "covered": False,
            "source": SOURCE["label"], "attribution": SOURCE["attribution"],
            "note": "This feed covers the Baltic and the Gulf of Finland. "
                    "There is no open global AIS service without an account.",
        }

    data = _all_positions()
    names = _all_names()

    centre_lon, centre_lat = (west + east) / 2, (south + north) / 2
    out = []
    for feature in data.get("features", []):
        lon, lat = (feature.get("geometry", {}).get("coordinates") or [None, None])[:2]
        if lon is None or not (west <= lon <= east and south <= lat <= north):
            continue
        props = feature.get("properties", {})
        mmsi = feature.get("mmsi") or props.get("mmsi")
        known = names.get(int(mmsi), {}) if mmsi is not None else {}
        category, type_label = _kind(known.get("type"))

        speed = props.get("sog")
        out.append({
            "mmsi": mmsi,
            "name": known.get("name"),
            "callsign": known.get("callsign"),
            "lon": lon, "lat": lat,
            # Speed over ground in knots; 102.3 is the "not available" value.
            "speed": None if speed is None or speed >= 102.0 else round(float(speed), 1),
            # Course over ground, and heading if the ship reports one. 511 is
            # the "no heading" value and must not be drawn as due north.
            "course": None if props.get("cog") is None or props["cog"] >= 360 else round(float(props["cog"]), 1),
            "heading": None if props.get("heading") is None or props["heading"] >= 511 else int(props["heading"]),
            "status": NAV_STATUS.get(props.get("navStat"), None),
            "category": category,
            "type": type_label,
            "destination": known.get("destination"),
            "draught": (known.get("draught") / 10.0) if known.get("draught") else None,
            "length": known.get("length"),
            "age_min": _age(props.get("timestampExternal") or props.get("timestamp")),
        })

    out.sort(key=lambda v: math.hypot(v["lon"] - centre_lon, v["lat"] - centre_lat))
    return {
        "vessels": out[:limit],
        "count": len(out),
        "covered": True,
        "source": SOURCE["label"],
        "attribution": SOURCE["attribution"],
        "fetched": dt.datetime.utcnow().isoformat(timespec="seconds") + "Z",
    }


# ── Synthetic vessels (DEMO_MODE) ──────────────────────────────


def demo_vessels(bbox: tuple[float, float, float, float], count: int = 40) -> dict:
    """Believable shipping for exploring offline.

    Laid along a couple of lanes rather than scattered, because scattered dots
    look nothing like traffic and would give a false impression of what the
    real layer shows.
    """
    west, south, east, north = bbox
    span_x, span_y = east - west, north - south
    out = []
    for i in range(count):
        t = (i + 0.5) / count
        lane = i % 3
        lon = west + span_x * t
        lat = south + span_y * (0.25 + 0.22 * lane + 0.05 * math.sin(t * 9 + lane))
        category, label = list(TYPES.values())[i % len(TYPES)]
        heading = int((90 if lane % 2 == 0 else 270) + 12 * math.sin(t * 7))
        out.append({
            "mmsi": 200000000 + i * 7717,
            "name": f"DEMO {chr(65 + i % 26)}{i:02d}",
            "callsign": None,
            "lon": lon, "lat": lat,
            "speed": round(2 + 16 * abs(math.sin(t * 5 + lane)), 1),
            "course": float(heading % 360),
            "heading": heading % 360,
            "status": "under way (engine)",
            "category": category, "type": label,
            "destination": None, "draught": None, "length": None,
            "age_min": round(0.5 + 3 * t, 1),
        })
    return {
        "vessels": out, "count": len(out), "covered": True, "demo": True,
        "source": "Synthetic", "attribution": "Synthetic demo data",
        "fetched": dt.datetime.utcnow().isoformat(timespec="seconds") + "Z",
    }
